### PR TITLE
Add IndexOf, EndsWith and StartWith

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/2922

### CODE REVIEW

**Comment by shargon** [comment]
@superboyiii could you test the gas cost?

File: src/Neo/SmartContract/Native/StdLib.cs [hunk 1]
```diff
@@ -240,5 +240,57 @@ private static int StrLen([MaxLength(MaxInputLength)] string str)
...
+            {
+                int k = 0;
+                for (; k < len; k++)
+                {
+                    if (byteToFind[k] != byteArray[i + k]) break;
+                }
+                if (k == len) return i;
+            }
+
+            return -1;
```
Comment (Liaojinghui) [inline src/Neo/SmartContract/Native/StdLib.cs:266]:
```suggestion
             return byteArray.AsSpan().IndexOf(byteToFind.AsSpan());
```

**Review by Liaojinghui — DISMISSED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Comment by shargon** [comment]
As @roman-khimov [said](https://github.com/neo-project/neo/pull/2926#pullrequestreview-1654434323):

> there is MemorySearch there. SUBSTR opcode is available as well.

Maybe we should remove IndexOf

**Comment by Liaojinghui** [comment]
name kinda confusing, but  you and @roman-khimov  are right.

**Comment by shargon** [comment]
Now I think that everything can be done with `MemorySearch`, these methods would only be synonyms of `MemorySearch`

**Comment by Liaojinghui** [comment]
> Now I think that everything can be done with `MemorySearch`, these methods would only be synonyms of `MemorySearch`

I also see that now. @roman-khimov  Roman is obviously right.

**Comment by shargon** [comment]
@Liaojinghui It should be done in neo-devpack-dotnet, using `MemorySearch`

**Comment by Liaojinghui** [comment]
> @Liaojinghui It should be done in neo-devpack-dotnet, using `MemorySearch`

Totally agree.

**Comment by vncoelho** [comment]
It can be synonyms of MemorySearch, maybe it is better.

How about the possible attacks in case of future resync, what is the idea? Mark all with hard fork?

**Comment by Liaojinghui** [comment]
@vncoelho This one will be replaced, new implementation will be in the devpack, and no fork will happen

**Comment by vncoelho** [comment]
Sure, just take care because of the other PR involving kecak

**Comment by shargon** [comment]
Closed because https://github.com/neo-project/neo-devpack-dotnet/pull/796

**Pull request closed without merging**